File: app/server/app/routers/opening.py
"""
Router para operaciones CRUD de Opening Counts
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
import uuid
from datetime import datetime

from app.database import get_db
from app.models import OpeningCount as OpeningCountModel
from app.schemas import OpeningCount, OpeningCountCreate

router = APIRouter()


@router.post("/", response_model=OpeningCount)
def create_opening_count(
    opening_count: OpeningCountCreate,
    db: Session = Depends(get_db)
):
    """Crear un nuevo conteo de apertura"""
    db_opening = OpeningCountModel(
        id=str(uuid.uuid4()),
        register_id=opening_count.register_id,
        store_id=opening_count.store_id,
        amount=opening_count.amount,
        date=opening_count.date,
        user_id=opening_count.user_id,
        user_name=opening_count.user_name,
        timestamp=datetime.now()
    )
    db.add(db_opening)
    db.commit()
    db.refresh(db_opening)
    return db_opening


@router.get("/", response_model=List[OpeningCount])
def get_opening_counts(
    skip: int = 0,
    limit: int = 100,
    store_id: str = None,
    date: str = None,
    db: Session = Depends(get_db)
):
    """Obtener todos los conteos de apertura"""
    query = db.query(OpeningCountModel)
    
    if store_id:
        query = query.filter(OpeningCountModel.store_id == store_id)
    if date:
        query = query.filter(OpeningCountModel.date == date)
    
    return query.offset(skip).limit(limit).all()


@router.get("/{opening_id}", response_model=OpeningCount)
def get_opening_count(opening_id: str, db: Session = Depends(get_db)):
    """Obtener un conteo de apertura por ID"""
    opening_count = db.query(OpeningCountModel).filter(
        OpeningCountModel.id == opening_id
    ).first()
    
    if not opening_count:
        raise HTTPException(status_code=404, detail="Opening count not found")
    
    return opening_count


@router.put("/{opening_id}", response_model=OpeningCount)
def update_opening_count(
    opening_id: str,
    opening_count: OpeningCountCreate,
    db: Session = Depends(get_db)
):
    """Actualizar un conteo de apertura"""
    db_opening = db.query(OpeningCountModel).filter(
        OpeningCountModel.id == opening_id
    ).first()
    
    if not db_opening:
        raise HTTPException(status_code=404, detail="Opening count not found")
    
    for key, value in opening_count.dict().items():
        setattr(db_opening, key, value)
    
    db.commit()
    db.refresh(db_opening)
    return db_opening


@router.delete("/{opening_id}")
def delete_opening_count(opening_id: str, db: Session = Depends(get_db)):
    """Eliminar un conteo de apertura"""
    opening_count = db.query(OpeningCountModel).filter(
        OpeningCountModel.id == opening_id
    ).first()
    
    if not opening_count:
        raise HTTPException(status_code=404, detail="Opening count not found")
    
    db.delete(opening_count)
    db.commit()
    return {"message": "Opening count deleted successfully"}

File: app/server/src/routes/opening.ts
import { Router, Request, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';
import { OpeningCount } from '../models/OpeningCount';
import { openingCountCreateSchema } from '../schemas/opening';

const router = Router();

// Crear un nuevo conteo de apertura
router.post('/', async (req: Request, res: Response) => {
  try {
    const validatedData = openingCountCreateSchema.parse(req.body);
    
    const openingCount = await OpeningCount.create({
      id: uuidv4(),
      ...validatedData,
      timestamp: new Date(),
    });

    res.status(201).json(openingCount);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ error: 'Datos inválidos', details: error.errors });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

// Obtener todos los conteos de apertura
router.get('/', async (req: Request, res: Response) => {
  try {
    const { skip = 0, limit = 100, store_id, date } = req.query;
    
    const where: any = {};
    if (store_id) where.store_id = store_id;
    if (date) where.date = date;

    const openingCounts = await OpeningCount.findAll({
      where,
      limit: Number(limit),
      offset: Number(skip),
    });

    res.json(openingCounts);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Obtener un conteo de apertura por ID
router.get('/:opening_id', async (req: Request, res: Response) => {
  try {
    const openingCount = await OpeningCount.findByPk(req.params.opening_id);
    
    if (!openingCount) {
      return res.status(404).json({ error: 'Opening count not found' });
    }

    res.json(openingCount);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Actualizar un conteo de apertura
router.put('/:opening_id', async (req: Request, res: Response) => {
  try {
    const validatedData = openingCountCreateSchema.parse(req.body);
    const openingCount = await OpeningCount.findByPk(req.params.opening_id);
    
    if (!openingCount) {
      return res.status(404).json({ error: 'Opening count not found' });
    }

    await openingCount.update(validatedData);
    res.json(openingCount);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ error: 'Datos inválidos', details: error.errors });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

// Eliminar un conteo de apertura
router.delete('/:opening_id', async (req: Request, res: Response) => {
  try {
    const openingCount = await OpeningCount.findByPk(req.params.opening_id);
    
    if (!openingCount) {
      return res.status(404).json({ error: 'Opening count not found' });
    }

    await openingCount.destroy();
    res.json({ message: 'Opening count deleted successfully' });
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

